File: desing-pattens-1/src/modelo/implementacao/desconto/DescontoVendaComLapisECaneta.java
package modelo.implementacao.desconto;

import interfaces.IDesconto;
import modelo.Orcamento;

public class DescontoVendaComLapisECaneta implements IDesconto {

	private IDesconto proximo;

	@Override
	public double calcula(Orcamento orcamento) {

		if (existeLapisOuCaneta(orcamento)) {
			return 0.05;
		}

		return proximo.calcula(orcamento);
	}

	private boolean existeLapisOuCaneta(Orcamento orcamento) {

		return buscaItem(orcamento, "lapis") && buscaItem(orcamento, "caneta");
	}

	private boolean buscaItem(Orcamento orcamento, String nomeItem) {

		return orcamento.getItens().stream().filter(item -> item.getNome().equalsIgnoreCase(nomeItem)).findAny().isPresent();
	}

	@Override
	public void setProximo(IDesconto desconto) {

		this.proximo = desconto;

	}

}

File: desing-pattens-1/src/main/java/modelo/implementacao/desconto/DescontoVendaComLapisECaneta.kt
package main.java.modelo.implementacao.desconto

import main.java.interfaces.IDesconto
import main.java.modelo.Orcamento

class DescontoVendaComLapisECaneta : IDesconto {

    private var proximo: IDesconto? = null

    override fun calcula(orcamento: Orcamento): Double {

        return if (existeLapisOuCaneta(orcamento)) {
            0.05
        } else proximo!!.calcula(orcamento)

    }

    private fun existeLapisOuCaneta(orcamento: Orcamento): Boolean {

        return buscaItem(orcamento, "lapis") && buscaItem(orcamento, "caneta")
    }

    private fun buscaItem(orcamento: Orcamento, nomeItem: String): Boolean {

        return orcamento.itens.stream().filter { item -> item.nome.equals(nomeItem, ignoreCase = true) }.findAny().isPresent
    }

    override fun setProximo(desconto: IDesconto) {

        this.proximo = desconto

    }

}
